doc/source/reference/config.rst
***********************
Configuration Settings 
***********************

In Lux, users can customize various global settings to configure the behavior of Lux through :py:class:`lux.config.Config`. These configurations are applied across all dataframes in the session. This page documents some of the configurations that you can apply in Lux.

.. contents:: :local:

.. note::

    Lux caches past generated recommendations, so if you have already printed the dataframe in the past, the recommendations would not be regenerated with the new config properties. In order for the config properties to apply, you would need to explicitly expire the recommendations as such:
        
        .. code-block:: python

            df = pd.read_csv("..")
            df # recommendations already generated here

            df.expire_recs()
            lux.config.SOME_SETTING = "..."
            df # recommendation will be generated again here

    Alternatively, you can place the config settings before you first print out the dataframe for the first time: 
        
        .. code-block:: python

            df = pd.read_csv("..")
            lux.config.SOME_SETTING = "..."
            df # recommendations generated for the first time with config



Change the default display 
~~~~~~~~~~~~~~~~~~~~~~~~~~~

We can set the :code:`default_display` to change whether the Pandas table or Lux widget is displayed by default. In the following block, we set the default display to 'lux', therefore the Lux widget will display first.

.. code-block:: python

    lux.config.default_display = "lux" 
    df

.. image:: https://github.com/lux-org/lux-resources/blob/master/doc_img/display-1.png?raw=true
  :width: 700
  :align: center

We can set the default_display back to 'pandas,' which would allow for the dataframe object to display first. You can still toggle to Lux/Pandas respectively using the 'Toggle' button.

.. code-block:: python

    lux.config.default_display = "pandas" # Set Pandas as default display
    df

.. image:: https://github.com/lux-org/lux-resources/blob/master/doc_img/display-2.png?raw=true
  :width: 700
  :align: center

If you try to set the default_display to anything other than 'lux' or 'pandas,' a warning will be shown, and the display will default to the previous setting.

.. code-block:: python
    
    lux.config.default_display = "notpandas" # Throw an warning
    df

.. image:: https://github.com/lux-org/lux-resources/blob/master/doc_img/display-3.png?raw=true
  :width: 700
  :align: center


Change the plotting scale 
~~~~~~~~~~~~~~~~~~~~~~~~~~~

We can adjust the sizes of charts displayed in the Lux widget based on a :code:`plotting_scale` applied to all charts. The default plotting scale is set to 1, displayed as followed: 

.. image:: https://github.com/lux-org/lux-resources/blob/master/doc_img/scale-0.png?raw=true
  :width: 700
  :align: center

In the following block, we set the :code:`plotting_scale` to 2. We see that the Lux widget displays charts two times larger than the default. 

.. code-block:: python

    lux.config.plotting_scale = 2 
    df

.. image:: https://github.com/lux-org/lux-resources/blob/master/doc_img/scale-1.png?raw=true
  :width: 700
  :align: center

When further operations are performed (e.g., specifying intent), the user-defined plotting scale continues to apply to both the current visualization and recommended charts. In addition to specifying :code:`plotting_scale` as a positive float (i.e., >1.0, enlarging the charts), :code:`plotting_scale` can also take in any positive float between [0,1] to shrink the size of the charts.


.. code-block:: python

    lux.config.plotting_scale = 0.5
    df

.. image:: https://github.com/lux-org/lux-resources/blob/master/doc_img/scale-3.png?raw=true
  :width: 700
  :align: center

The :code:`plotting_scale` config takes in any positive float or integer and functions for both Altair and Matplotlib plotting backends.



Change plotting backend for rendering visualizations in Lux
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

We can set the :code:`plotting_backend` config to change the plotting library used for rendering the visualizations in Lux. 
This is often useful not just for stylizing plot aesthetics, but also to change the code generated when `exporting a visualization <https://lux-api.readthedocs.io/en/latest/source/guide/export.html>`__.
For example, if you are more familiar with `matplotlib <https://matplotlib.org/>`__ , you may want to use a matplotlib plotting backend so that you can make use of the exported visualization code. In the following code, we set the plotting backend to 'matplotlib', and Lux will display the Matplotlib rendered charts.

.. code-block:: python

    lux.config.plotting_backend = "matplotlib" 
    df

.. image:: https://github.com/lux-org/lux-resources/blob/master/doc_img/vislib-1.png?raw=true
  :width: 700
  :align: center

We can set the vislib back to the default 'vegalite,' which uses Vega-Lite to render the displayed chart.

.. code-block:: python

    lux.config.plotting_backend = "vegalite" 
    df

.. image:: https://github.com/lux-org/lux-resources/blob/master/doc_img/display-1.png?raw=true
  :width: 700
  :align: center

Lux currently only support Vega-Lite and matplotlib, and we plan to add support for other plotting libraries in the future. If you try to set the :code:`plotting_backend` to anything other than 'matplotlib' or 'vegalite', a warning will be shown, and the display will default to the previous setting.

.. code-block:: python
    
    lux.config.plotting_backend = "notvegalite" # Throw an warning
    df

.. image:: https://github.com/lux-org/lux-resources/blob/master/doc_img/vislib-2.png?raw=true

  :width: 700
  :align: center

Change the sampling parameters 
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To speed up the visualization processing, by default, Lux performs random sampling on datasets with more than 100000 rows. Specifically, for datasets over 100000 rows, Lux will randomly sample 100000 rows from the dataset.

If we want to change these parameters, we can set the `sampling_thresh` via `lux.config` to change the default form of output. The `sampling_thresh` is by default set to 100000. In the following block, we increase this sampling threshold.

.. code-block:: python

    lux.config.sampling_thresh = 500000

If we want Lux to use the full dataset in the visualization, we can also disable sampling altogether (but note that this may result in long processing times). Below is an example of disabling the sampling:

.. code-block:: python

    lux.config.sampling = False

Disable the use of heatmaps for large datasets
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In addition to sampling, Lux replaces scatter plots with heatmaps for datasets with over 5000 rows to speed up the visualization process.

We can disable this feature and revert back to using a scatter plot by running the following code block (but note that this may result in long processing times).

.. code-block:: python

    lux.config.heatmap = False

Changing the plotting style
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In Lux, we can change the chart settings and aesthetics by inputting global custom plot settings the :code:`plotting_style`.
For charts rendered in Altair (default), we can change the plotting style by writing a function that takes a `AltairChart <https://altair-viz.github.io/user_guide/generated/toplevel/altair.Chart.html>`_ object as input and output. For example:

.. code-block:: python

    def change_color_add_title(chart):
        chart = chart.configure_mark(color="green") # change mark color to green
        chart.title = "Custom Title" # add title to chart
        return chart

Then, set the `plotting_style` to this function so that this function is applied to every plot generated.

.. code-block:: python

    lux.config.plotting_style = change_color_add_title

The above results in the following changes:

.. image:: https://github.com/lux-org/lux-resources/blob/master/doc_img/style-2.png?raw=true
  :width: 600
  :align: center

Matplotlib also supports plot configurations to be applied on top of the generated graphs. To set a default plot configuration, first write a function that can take in a `fig` and `ax` and returns a `fig` and `ax`.
`fig` handles figure width and other plot size attributes. `ax` supports changing the chart title and other plot labels and configurations. For example:

.. code-block:: python
	
    lux.config.plotting_backend = "matplotlib"

    def change_width_add_title(fig, ax):
        fig.set_figwidth(7) # change figure width
        ax.set_title("Custom Title") # add title to chart
        return fig, ax

    lux.config.plotting_style = change_width_add_title

See `this page <https://lux-api.readthedocs.io/en/latest/source/guide/style.html>`__ for more details.


Modify Sorting and Ranking in Recommendations
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In Lux, we select a small subset of visualizations to display in each action tab to avoid displaying too many charts at once. 
Certain recommendation categories ranks and selects the top K most interesting visualizations to display.
You can modify the sorting order and selection cutoff via :code:`lux.config`.
By default, the recommendations are sorted in a :code:`"descending"` order based on their interestingness score, you can reverse the ordering by setting the sort order as:

.. code-block:: python 

    lux.config.sort = "ascending"

To turn off the sorting of visualizations based on its score completely and ensure that the visualizations show up in the same order across all dataframes, you can set the sorting as "none":

.. code-block:: python 

    lux.config.sort = "none"

For recommendation actions that generate a lot of visualizations, we select the cutoff criteria as the top 15 visualizations. If you would like to see only see the top 6 visualizations, you can set:

.. code-block:: python 

    lux.config.topk = 6

If you would like to turn off the selection criteria completely and display everything, you can turn off the top K selection by:

.. code-block:: python 

    lux.config.topk = False

Beware that this may generate large numbers of visualizations (e.g., for 10 quantitative variables, this will generate 45 scatterplots in the Correlation action!)

Changing heatmap bin resolution
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In Lux, large scatterplots are displayed as heatmaps that are 40x40 by default. You can increase or decrease the heatmap resolution NxN by changing the bin size N:

.. code-block:: python 

    lux.config.heatmap_bin_size = 100

This generates heatmap visualizations that are binned into a 100x100 grid. 



Change the maximum number of bars displayed
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

In Lux, we limit the maximum number of bars displayed in a bar chart to 10 bars to avoid cluttering. 
If you want Lux to display more bars in each chart, we can set the :code:`number_of_bars` to increase the maximum number of bars displayed. 

.. code-block:: python

    lux.config.number_of_bars = 20

.. image:: https://github.com/lux-org/lux-resources/blob/master/doc_img/config_bars.png?raw=true
  :width: 400
  :align: left


doc/source/reference/lux._config.config.rst
lux.config.config package
===================================

lux._config.config.Config module
---------------------------------

.. automodule:: lux._config.config.config
    :members:



lux/_config/config.py
"""
This config file was largely borrowed from Pandas config.py set_action functionality.
For more resources, see https://github.com/pandas-dev/pandas/blob/master/pandas/_config
"""
from collections import namedtuple
from typing import Any, Callable, Dict, Iterable, List, Optional, Union
import lux
import warnings
from lux.utils.tracing_utils import LuxTracer
import os
from lux._config.template import postgres_template, mysql_template

RegisteredOption = namedtuple("RegisteredOption", "name action display_condition args")


class Config:
    """
    Class for Lux configurations applied globally across entire session
    """

    def __init__(self):
        self._default_display = "pandas"
        self.plotting_style = None
        self.SQLconnection = ""
        self.executor = None
        # holds registered option metadata
        self.actions: Dict[str, RegisteredOption] = {}
        # flags whether or not an action has been registered or removed and should be re-rendered by frame.py
        self.update_actions: Dict[str, bool] = {}
        self.update_actions["flag"] = False
        self._plotting_backend = "vegalite"
        self._plotting_scale = 1
        self._topk = 15
        self._number_of_bars = 10  # max no of bars displayed (rest shown as "+ k more")
        self._label_len = 25  # max length of x and y axis labels
        self._sort = "descending"
        self._pandas_fallback = True
        self._interestingness_fallback = True
        self.heatmap_bin_size = 40
        self.tracer_relevant_lines = []
        self.tracer = LuxTracer()
        self.query_templates = {}
        self.handle_quotes = True
        #####################################
        #### Optimization Configurations ####
        #####################################
        self._sampling_thresh = 100000
        self._sampling_flag = True
        self._heatmap_flag = True
        self._heatmap_start = 5000
        self.lazy_maintain = True
        self.early_pruning = True
        self.early_pruning_sample_cap = 30000
        # Apply sampling only if the dataset is 150% larger than the sample cap
        self.early_pruning_sample_start = self.early_pruning_sample_cap * 1.5
        self.streaming = False
        self.render_widget = True

    @property
    def number_of_bars(self):
        return self._number_of_bars

    @number_of_bars.setter
    def number_of_bars(self, k: int) -> None:
        """
        Parameters
        ----------
        k : int
            Number of bars in output bar charts; rest are not displayed
        """
        if type(k) == int:
            self._number_of_bars = k
        else:
            warnings.warn(
                "The number of bars must be an integer.",
                stacklevel=2,
            )

    @property
    def label_len(self):
        return self._label_len

    @label_len.setter
    def label_len(self, l: int) -> None:
        """
        Parameters
        ----------
        l : int
            Maximum length of string axis labels
        """
        if type(l) == int:
            self._label_len = l
        else:
            warnings.warn(
                "The maximum length must be an integer.",
                stacklevel=2,
            )

    @property
    def topk(self):
        return self._topk

    @topk.setter
    def topk(self, k: Union[int, bool]):
        """
        Setting parameter to display top k visualizations in each action

        Parameters
        ----------
        k : Union[int,bool]
            False: if display all visualizations (no top-k)
            k: number of visualizations to display
        """
        if isinstance(k, int) or isinstance(k, bool):
            self._topk = k
        else:
            warnings.warn(
                "Parameter to lux.config.topk must be an integer or a boolean.",
                stacklevel=2,
            )

    @property
    def sort(self):
        return self._sort

    @sort.setter
    def sort(self, flag: Union[str]):
        """
        Setting parameter to determine sort order of each action

        Parameters
        ----------
        flag : Union[str]
            "none", "ascending","descending"
            No sorting, sort by ascending order, sort by descending order
        """
        flag = flag.lower()
        if isinstance(flag, str) and flag in ["none", "ascending", "descending"]:
            self._sort = flag
        else:
            warnings.warn(
                "Parameter to lux.config.sort must be one of the following: 'none', 'ascending', or 'descending'.",
                stacklevel=2,
            )

    @property
    def pandas_fallback(self):
        return self._pandas_fallback

    @pandas_fallback.setter
    def pandas_fallback(self, fallback: bool) -> None:
        """
        Parameters
        ----------
        fallback : bool
            If an error occurs, whether or not to raise an exception or fallback to default Pandas.
        """
        if type(fallback) == bool:
            self._pandas_fallback = fallback
        else:
            warnings.warn(
                "The flag for Pandas fallback must be a boolean.",
                stacklevel=2,
            )

    @property
    def interestingness_fallback(self):
        return self._interestingness_fallback

    @interestingness_fallback.setter
    def interestingness_fallback(self, fallback: bool) -> None:
        """
        Parameters
        ----------
        fallback : bool
            If an error occurs while calculating interestingness, whether or not
            to raise an exception or fallback to default Pandas.
        """
        if type(fallback) == bool:
            self._interestingness_fallback = fallback
        else:
            warnings.warn(
                "The flag for interestingness fallback must be a boolean.",
                stacklevel=2,
            )

    @property
    def sampling_thresh(self):
        return self._sampling_thresh

    @sampling_thresh.setter
    def sampling_thresh(self, sample_number: int) -> None:
        """
        Parameters
        ----------
        sample_number : int
            Number of rows required to begin sampling.

        """
        if type(sample_number) == int:
            self._sampling_thresh = sample_number
        else:
            warnings.warn(
                "The sampling starting point must be an integer.",
                stacklevel=2,
            )

    @property
    def sampling(self):
        """
        Parameters
        ----------
        sample_flag : bool
            Whether or not sampling will occur.
        """
        return self._sampling_flag

    @sampling.setter
    def sampling(self, sample_flag: bool) -> None:
        """
        Parameters
        ----------
        sample_flag : bool
            Whether or not sampling will occur.
        """
        if type(sample_flag) == bool:
            self._sampling_flag = sample_flag
        else:
            warnings.warn(
                "The flag for sampling must be a boolean.",
                stacklevel=2,
            )

    @property
    def heatmap(self):
        """
        Parameters
        ----------
        heatmap_flag : bool
            Whether or not a heatmap will be used instead of a scatter plot.
        """
        return self._heatmap_flag

    @heatmap.setter
    def heatmap(self, heatmap_flag: bool) -> None:
        """
        Parameters
        ----------
        heatmap_flag : bool
            Whether or not a heatmap will be used instead of a scatter plot.
        """
        if type(heatmap_flag) == bool:
            self._heatmap_flag = heatmap_flag
        else:
            warnings.warn(
                "The flag for enabling/disabling heatmaps must be a boolean.",
                stacklevel=2,
            )

    @property
    def default_display(self):
        """
        Set the widget display to show Pandas by default or Lux by default
        Parameters
        ----------
        type : str
            Default display type, can take either the string `lux` or `pandas` (regardless of capitalization)
        """
        return self._default_display

    @default_display.setter
    def default_display(self, type: str) -> None:
        """
        Set the widget display to show Pandas by default or Lux by default
        Parameters
        ----------
        type : str
            Default display type, can take either the string `lux` or `pandas` (regardless of capitalization)
        """
        if type.lower() == "lux":
            self._default_display = "lux"
        elif type.lower() == "pandas":
            self._default_display = "pandas"
        else:
            warnings.warn(
                "Unsupported display type. Default display option should either be `lux` or `pandas`.",
                stacklevel=2,
            )

    @property
    def plotting_backend(self):
        return self._plotting_backend

    @plotting_backend.setter
    def plotting_backend(self, type: str) -> None:
        """
        Set the widget display to show Vegalite by default or Matplotlib by default
        Parameters
        ----------
        type : str
                Default display type, can take either the string `vegalite` or `matplotlib` (regardless of capitalization)
        """
        if type.lower() == "vegalite" or type.lower() == "altair":
            self._plotting_backend = "vegalite"
        elif type.lower() == "matplotlib":
            self._plotting_backend = "matplotlib_svg"
        else:
            warnings.warn(
                "Unsupported plotting backend. Lux currently only support 'altair', 'vegalite', or 'matplotlib'",
                stacklevel=2,
            )

    @property
    def plotting_scale(self):
        return self._plotting_scale

    @plotting_scale.setter
    def plotting_scale(self, scale: float) -> None:
        """
        Set the scale factor for charts displayed in Lux.
        ----------
        type : float (default = 1.0)
        """
        scale = float(scale) if isinstance(scale, int) else scale
        if isinstance(scale, float) and scale > 0:
            self._plotting_scale = scale
        else:
            warnings.warn(
                "Scaling factor for charts must be a positive float.",
                stacklevel=2,
            )

    def _get_action(self, pat: str, silent: bool = False):
        return lux.actions[pat]

    def register_action(
        self,
        name: str = "",
        action: Callable[[Any], Any] = None,
        display_condition: Optional[Callable[[Any], Any]] = None,
        *args,
    ) -> None:
        """
        Registers the provided action globally in lux

        Parameters
        ----------
        name : str
                the name of the action
        action : Callable[[Any], Any]
                the function used to generate the recommendations
        display_condition : Callable[[Any], Any]
                the function to check whether or not the function should be applied
        args: Any
                any additional arguments the function may require
        """
        if action:
            if not callable(action):
                raise ValueError("Action must be a callable")
        if display_condition:
            if not callable(display_condition):
                raise ValueError("Display condition must be a callable")
        self.actions[name] = RegisteredOption(
            name=name, action=action, display_condition=display_condition, args=args
        )
        self.update_actions["flag"] = True

    def remove_action(self, name: str = "") -> None:
        """
        Removes the provided action globally in lux

        Parameters
        ----------
        name : str
                the name of the action to remove
        """
        if name not in self.actions:
            raise ValueError(f"Option '{name}' has not been registered")

        del self.actions[name]
        self.update_actions["flag"] = True

    def set_SQL_connection(self, connection):
        """
        Sets SQL connection to a database

        Parameters:
            connection : SQLAlchemy connectable, str, or sqlite3 connection
                For more information, `see here <https://docs.sqlalchemy.org/en/13/core/connections.html>`__
        """
        self.set_executor_type("SQL")
        self.SQLconnection = connection

    def read_query_template(self, query_template):
        from lux.executor.SQLExecutor import SQLExecutor

        query_dict = {}
        if type(query_template) is str:
            for line in query_template.split("\n"):
                (key, val) = line.split(":")
                query_dict[key] = val.strip()
        else:
            with open(query_file) as f:
                for line in f:
                    (key, val) = line.split(":")
                    query_dict[key] = val.strip()
        self.query_templates = query_dict
        self.executor = SQLExecutor()

    def set_executor_type(self, exe):
        if exe == "SQL":
            from lux.executor.SQLExecutor import SQLExecutor

            self.executor = SQLExecutor()
            self.read_query_template(postgres_template)
        elif exe == "Pandas":
            from lux.executor.PandasExecutor import PandasExecutor

            self.SQLconnection = ""
            self.executor = PandasExecutor()
        else:
            raise ValueError("Executor type must be either 'Pandas' or 'SQL'")


def warning_format(message, category, filename, lineno, file=None, line=None):
    return "%s:%s: %s:%s\n" % (filename, lineno, category.__name__, message)
